qiskit/circuit/library/grover_operator.py

```diff
@@ -0,0 +1,235 @@
+        Args:
+            oracle: The oracle implementing a reflection about the bad state.
+            state_in: The operator preparing the good and bad state. For Grover's algorithm,
+                this is a n-qubit Hadamard gate and for Amplitude Amplification or Estimation
+                the operator A.
+            zero_reflection: The reflection about the zero state.
+            reflection_qubits: Qubits on which the zero reflection acts on.
+            insert_barriers: Whether barriers should be inserted between the reflections and A.
+            mcx_mode: The mode to use for building the default zero reflection.
+            name: The name of the circuit.
+        """
+        super().__init__(name=name)
+
+        # store inputs
+        if isinstance(oracle, Statevector):
+            from qiskit.circuit.library import Diagonal  # pylint: disable=cyclic-import
+            oracle = Diagonal((-1) ** oracle.data)
+        self._oracle = oracle
+
+        if isinstance(zero_reflection, (Operator, DensityMatrix)):
+            from qiskit.circuit.library import Diagonal  # pylint: disable=cyclic-import
+            zero_reflection = Diagonal(zero_reflection.data.diagonal())
+        self._zero_reflection = zero_reflection
+
+        self._reflection_qubits = reflection_qubits
+        self._state_in = state_in
+        self._insert_barriers = insert_barriers
+        self._mcx_mode = mcx_mode
+
+        # build circuit
+        self._build()
+
+    @property
+    def reflection_qubits(self):
+        """Reflection qubits, on which S0 is applied (if S0 is not user-specified)."""
+        if self._reflection_qubits is not None:
+            return self._reflection_qubits
+
+        num_state_qubits = self.oracle.num_qubits - self.oracle.num_ancillas
+        return list(range(num_state_qubits))
+
+    @property
+    def zero_reflection(self) -> QuantumCircuit:
+        """The subcircuit implementing the reflection about 0."""
+        if self._zero_reflection is not None:
+            return self._zero_reflection
+
+        num_state_qubits = self.oracle.num_qubits - self.oracle.num_ancillas
+        qubits = self.reflection_qubits
+        return _zero_reflection(num_state_qubits, qubits, self._mcx_mode)
+
+    @property
+    def state_in(self) -> QuantumCircuit:
+        """The subcircuit implementing the A operator or Hadamards."""
+        if self._state_in is not None:
+            return self._state_in
+
+        num_state_qubits = self.oracle.num_qubits - self.oracle.num_ancillas
+        hadamards = QuantumCircuit(num_state_qubits, name='H')
+        # apply Hadamards only on reflection qubits, rest will cancel out
+        hadamards.h(self.reflection_qubits)
+        return hadamards
+
+    @property
+    def oracle(self):
+        """The oracle implementing a reflection about the bad state."""
+        return self._oracle
+
+    def _build(self):
+        num_state_qubits = self.oracle.num_qubits - self.oracle.num_ancillas
```

Comment:
At the moment, you cannot give the oracle as a `Gate` object (which would be cool) because `Gate`s don't have a `.num_ancillas`. I'm not sure what the best approach is, maybe adding the `.num_ancillas` method to the `Gate` class, or to catch the case and `0` if the oracle is a `Gate`? I'm unsure what the ramifications of either of these would be.